ufo2fontir/src/source.rs

```diff
@@ -877,47 +929,64 @@ impl Work<Context, WorkError> for KerningWork {
+            &static_metadata.glyph_order,
+            default_master,
+        )?;
+
+        // Group names are side-specific (public.kern1/2) but the set of glyph names isn't
+        // so include side in the key to avoid matching the wrong side
+        let reverse_groups: HashMap<_, _> = kerning
+            .groups
+            .iter()
+            .map(|(k, v)| ((KernSide::of(k), v), k))
+            .collect();
+
+        // https://gist.github.com/madig/76567a9650de639bbff51ce010783790#file-align-groups-py-L21
+        // claims some sources use different names for groups of the same glyphs in different masters
+        // so we need to create a renaming map.
+        let mut old_to_new_group_name: HashMap<_, _> = kerning
+            .groups
+            .keys()
+            .map(|name| (name.clone(), name))
+            .collect();
+
+        for (_, source) in self
             .designspace
             .sources
             .iter()
-            .filter(|s| !is_glyph_only(s))
+            .enumerate()
+            .filter(|(idx, source)| !is_glyph_only(source) && *idx != default_master_idx)
         {
-            let ufo_dir = designspace_dir.join(&source.filename);
-            let data_request = norad::DataRequest::none().groups(true);
-            let font = norad::Font::load_requested_data(&ufo_dir, data_request)
-                .map_err(|e| WorkError::ParseError(ufo_dir, format!("{e}")))?;
-
-            kerning.groups = font
-                .groups
-                .into_iter()
-                .filter(|(group_name, entries)|
-                    (group_name.starts_with(UFO_KERN1_PREFIX) || group_name.starts_with(UFO_KERN2_PREFIX))
-                    && !entries.is_empty()
-                )
-                .map(|(group_name, entries)| {
-                    (
-                        GroupName::from(group_name.as_str()),
-                        entries
-                            .into_iter()
-                            .filter_map(|glyph_name| {
-                                let glyph_name = GlyphName::from(glyph_name.as_str());
-                                if static_metadata.glyph_order.contains(&glyph_name) {
-                                    Some(glyph_name)
-                                } else {
-                                    debug!("{} kerning group '{}' references non-existent glyph '{}'; ignoring", source.filename, group_name, glyph_name);
-                                    None
-                                }
-                            })
-                            .collect(),
-                    )
-                })
-                .collect();
+            for (name, entries) in
+                groups_for(designspace_dir, &static_metadata.glyph_order, source)?
+            {
+                let Some(real_name) = reverse_groups.get(&(KernSide::of(&name), &entries)) else {
+                    trace!("{name} exists only in {} and will be ignored", source.name);
```

Comment:
I suggest you turn this to a warning level message because, in my understanding, non-default masters aren't supposed to define different groups from the default

Reply:
Done, ty